validphys2/src/validphys/commondataparser.py

```diff
@@ -1,18 +1,310 @@
+    ret = op_str.upper()
+    # TODO: move accepted operations to this module so that the convolution receives an operation to apply
+    # instead of an operation to understand
+    from validphys.convolution import OP
+
+    if ret not in OP:
+        raise ValidationError(f"The operation '{op_str}' is not implemented in validphys")
+    return ret
+
+
+# Auxiliary objects
+
+
+@dataclasses.dataclass
+class ValidApfelComb:
+    """Some of the grids might have been converted from apfelcomb and introduce hacks.
+    These are the allowed hacks:
+        - repetition_flag:
+            list of fktables which might need to be repeated
+            necessary to apply c-factors in compound observables
+        - normalization:
+            mapping with the single fktables which need to be normalized and the factor
+            note that when they are global factors they are promoted to conversion_factor
+        - shifts:
+            mapping with the single fktables and their respective shifts
+            necessary to create "gaps" that some cfactors or datasets might expect
+    """
+
+    repetition_flag: Optional[list[str]] = None
+    normalization: Optional[dict] = None
+    shifts: Optional[dict] = None
+
+
+@dataclasses.dataclass
+class TheoryMeta:
+    """Contains the necessary information to load the associated fktables"""
+
+    FK_tables: list[list]
+    operation: ValidOperation
+    conversion_factor: float = 1.0
+    comment: Optional[str] = None
+    apfelcomb: Optional[ValidApfelComb] = None
+    # The following options are transitional so that the yamldb can be used from the theory
+    appl: Optional[bool] = False
+    target_dataset: Optional[str] = None
+
+    def fktables_to_paths(self, grids_folder):
+        """Given a source for pineappl grids, constructs the lists of fktables
+        to be loaded"""
+        ret = []
+        for operand in self.FK_tables:
+            ret.append([grids_folder / f"{m}.{EXT}" for m in operand])
+        return ret
+
+    @classmethod
+    def parser(cls, meta: dict):
+        """The yaml databases in the server use "operands" instead of "FK_tables" """
+        if "operands" in meta:
+            meta["FK_tables"] = meta.pop("operands")
+        return parse_input(meta, cls)
+
+
+ValidTheory = Parser(TheoryMeta.parser)
+
+
+@dataclasses.dataclass
+class ValidKinematics:
+    """Contains all metadata for the kinematics of the dataset"""
+
+    file: ValidPath
+    variables: dict
+
+
+@dataclasses.dataclass
+class ValidReference:
+    """Holds literature information for the dataset"""
+
+    url: str
+    version: int = 0
+    tables: list[int] = dataclasses.field(default_factory=list)
+
+
+@dataclasses.dataclass
+class Variant:
+    """Defines the keys of the CommonMetaData that can be overwritten"""
+
+    data_uncertainties: list[ValidPath]
+
+
+@Parser
+def ValidVariants(variant_dict: dict) -> Dict[str, Variant]:
```

Comment:
This is not needed. Can be just the return type and will be processed recursively as done manually in the function.

Reply:
You mean that it will be able to understand that it is a `dict` of `Variants` by itself or the `Parser` is not needed?

Reply:
Yes it should.